### ISSUE TITLE
UI Design

### ISSUE THREAD

**Issue by seisatsu** [seisatsu/Driftwood#27]
Support UI design using <s>some sort of JSON descriptor language, or possibly CSS/HTML</s> python.

**Comment by seisatsu** [comment]
We currently support containers and text.

**Comment by pmer** [comment]
Clickable buttons is a good place to start with this for Alpha-0.0.4, then nicer buttons (keyboard navigable, unfocused/focused states, inactive/active states) can be shipped after.

**Comment by pmer** [comment]
Will Driftwood support a mouse on desktop? Or are buttons & menus keyboard navigable?

**Comment by seisatsu** [comment]
I was thinking to do keyboard controlled UIs first, and I have a vague idea of how to do so.

**Comment by seisatsu** [comment]
With containers and text widgets working, the rest of the UI can be implemented in the STDlib.

**Comment by seisatsu** [comment]
This issue will be for making UIs possible in WidgetManager (keyboard, mouse support, etc.) while #135 will be for menu implementation.

**Comment by seisatsu** [comment]
Currently, there is no reason I can see to have a special "button" or "menu" widget. Any widget currently that isn't text is a container widget, and these don't necessarily need to contain anything. Usually though they will contain an image and/or some text. A container widget should be able to have an event attached to it and be used as a button. They can also contain multiple buttons. Radio buttons and check boxes can also be implemented as simple buttons, aka containers. Scrollbars can be done with containers and an image that changes places.

Basically I am thinking right now that I want WidgetManager to only handle container and text widgets and connecting keyboard/mouse input to them, routing events from widgets, etc. Everything else will be handled by other systems built on top of WidgetManager's basic functionality. #135 will be a comprehensive system for automatically building menus, but there will also be a ui standard library that automates making common but more complex widgets out of the two base widgets, and menu will utilize this.

**Comment by pmer** [comment]
Hmm, IIRC container widgets only support background images and don't support background fill or gradiants or box shadows. I have used such properties when making buttons for web apps in the past. (E.g., GitHub buttons use a gradiant & the buttons at the bottom of YouTube's homepage use a fill and box shadow.) There's no urgent need for it right now but they might be useful capabilities to develop sometime later.

**Comment by seisatsu** [comment]
I'm going to split this up.
